### PR TITLE
[PackageGraph] Add incomplete mode to resolver

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
This mode lets the resolver run without fetching the containers which
it doesn't already have. This is useful when we want to run the resolver
with different sets of constraints without going off cloning a lot of
new repositories which might be unnecessary.

This, with delta debugging algo, will eventually let us diagnose the
unresolvable package graphs.

### CODE REVIEW

File: Sources/PackageGraph/DependencyResolver.swift [hunk 1]
```diff
@@ -786,6 +786,16 @@ public class DependencyResolver<
...
 
+    /// Puts the resolver in incomplete mode.
+    ///
+    /// In this mode, no new containers will be requested from the provider.
+    /// Instead, if a container is not already present in the resolver, it will
+    /// skipped without raising an error. This is useful to avoid cloning
+    /// repositories from network when trying to partially resolve the constraints.
+    ///
+    /// Note that the input constraints will always be fetched.
+    public var incompleteMode = false
```
Comment (ddunbar) [inline Sources/PackageGraph/DependencyResolver.swift]: s/var/let/?

File: Tests/PackageGraphTests/DependencyResolverTests.swift [hunk 2]
```diff
@@ -551,6 +551,70 @@ class DependencyResolverTests: XCTestCase {
...
+                    (container: "B", requirement: .versionSet(v1Range)),
+                ],
+            ]),
+            MockPackageContainer(name: "B", dependencies: [
+                "1.0.0": [],
+            ]),
+        ])
+
+        let resolver = MockDependencyResolver(provider, MockResolverDelegate())
+        resolver.incompleteMode = true
```
Comment (ddunbar) [inline Tests/PackageGraphTests/DependencyResolverTests.swift]: Oh... would it be better to just make this a parameter and not use mutability, or make it an option on the resolve call?

Comment (aciidb0mb3r) [inline Tests/PackageGraphTests/DependencyResolverTests.swift]: I was thinking to have a parameter "debug if unsatisfiable" on the resolve() method. That will initially run the resolver in full mode and then put it in incomplete mode and start the delta debugging algo. The advantage is that the clients who don't want debugging won't get it (like the tests which check for failures). The idea was that the resolver is expected to run in normal mode before getting sent into incomplete mode but I think it should be fine if this is internal set for now. Does that make sense?

Comment (ddunbar) [inline Tests/PackageGraphTests/DependencyResolverTests.swift]: I see... yeah that seems fine

**Pull request merged**